capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,483 @@
+            {
+                rule: {
+                    uid: r
+                    for uid, r in res.items()
+                    if r.category == category and r
+                }
+                for rule, res in self.items()
+            }
+        )
+
+    def by_value(self, value: bool) -> Results:
+        """Filter the validation results by ``value``."""
+        return Results(
+            {
+                rule: {
+                    uid: r for uid, r in res.items() if r.value == value and r
+                }
+                for rule, res in self.items()
+            }
+        )
+
+    def by_type(self, type: str) -> Results:
+        """Filter the validation results by ``type``."""
+        results = dict[Rule, dict[helpers.UUIDString, Result]]()
+        for rule, res in self.items():
+            if not res:
+                continue
+
+            for rule_type in rule.types:
+                if rule_type != type:
+                    continue
+
+                results[rule] = res
+
+        return Results(results)
+
+    def setdefault(
+        self,
+        key: t.Any,
+        default: dict[helpers.UUIDString, Result] | None = None,
+    ) -> dict[helpers.UUIDString, Result]:
+        if not isinstance(default, dict):
+            raise ValueError("Only a result dictionary is accepted")
+        return super().setdefault(key, default)
+
+
+def register_rule(
+    category: Category,
+    types: (
+        str
+        | VirtualType[_T]
+        | type[_T]
+        | cabc.Iterable[str | VirtualType[_T] | type[_T]]
+    ),
+    id: str,
+    name: str,
+    rationale: str,
+    action: str,
+    hyperlink_further_reading: str | None = None,
+) -> cabc.Callable[[cabc.Callable[[_T], bool]], cabc.Callable[[_T], bool]]:
+    """Register the validation rule.
+
+    The decorator registers the validation rule function (validator)
+    along with the object type to be validated (type). The type is the
+    object type the rule is applicable to. The type will be used to feed
+    the validator with inputs (instances of type) during validation.
+    """
+    try:
+        VALIDATION_RULES.by_id(id)
+        LOGGER.warning(
+            "Failed to register rule: %r with ID: %r. Rule ID already exists",
+            name,
+            id,
+        )
+        unique_check = False
+    except KeyError:
+        unique_check = True
+
+    if not types:
+        raise TypeError("No 'types' specified")
+    if isinstance(types, type):
+        type_names = [types.__name__]
+    elif isinstance(types, (RealType, VirtualType)):
+        type_names = [types.name]
+    elif isinstance(types, str):
+        type_names = [types]
+    else:
+        type_names = []
+        for i in types:
+            if isinstance(i, str):
+                type_names.append(i)
+            elif isinstance(i, (RealType, VirtualType)):
+                type_names.append(i.name)
+            else:
+                type_names.append(i.__name__)
```

Comment:
`types` should end up being a tuple, to prevent it accidentally being modified later.